### DISCUSSION TITLE
How to get the body of a function that returns impl Responder ?

### DISCUSSION THREAD

**Discussion by alkeryn** [actix/actix-web#3005] [Q&A] [answered]
i tried something like that
```rust
        let resp = super::near(actix_web::web::Data::new(db), f, auth::User::default() ).await;
        let resp = resp.respond_to(&req);
        println!("{:?}", resp.body());
```
however the compiler tels me :
```
`<impl Responder as Responder>::Body` cannot be formatted using `{:?}` because it doesn't implement `Debug`
```

**Comment by robjtede** [comment, marked as answer]
[`body::to_bytes`](https://docs.rs/actix-web/latest/actix_web/body/fn.to_bytes.html)

**Comment by Samielakkad** [comment]
The body is not printable because it is still a body type/stream, not necessarily bytes in memory. Convert the response into an `HttpResponse`, take ownership of its body, then collect it with `body::to_bytes`.

In tests, the easier route is usually `test::call_service` or `test::call_and_read_body`, but if you already have a responder value:

```rust
use actix_web::{body, test, Responder};

let req = test::TestRequest::default().to_http_request();

let responder = super::near(
    actix_web::web::Data::new(db),
    f,
    auth::User::default(),
).await;

let response = responder.respond_to(&req);
let bytes = body::to_bytes(response.into_body()).await.unwrap();

println!("{}", String::from_utf8_lossy(&bytes));
```

If your handler returns a `Result<impl Responder, Error>`, unwrap/handle the `Result` first, then call `respond_to` on the responder.

The key idea is:

```text
impl Responder -> HttpResponse -> body -> bytes -> string/json
```

`Debug` on the body itself is not the right operation because the body may be streaming.
